### PR TITLE
perf(store): add `outsideZone` option to the config

### PR DESCRIPTION
**Pull request by arturovt**
## PR Checklist
- [x] The commit message follows our guidelines: https://github.com/ngxs/store/blob/master/CONTRIBUTING.md#commit
- [x] Tests for the changes have been added (for bug fixes / features)
- [x] Docs have been added / updated (for bug fixes / features)

## PR Type
What kind of change does this PR introduce?
```
[ ] Bugfix
[x] Feature
[x] Perfomance issue
[ ] Code style update (formatting, local variables)
[ ] Refactoring (no functional changes, no api changes)
[ ] Build related changes
[ ] CI related changes
[ ] Documentation content changes
[ ] Other... Please describe:
```

## Current behavior

https://github.com/ngxs/store/issues/739
https://github.com/ngxs/store/issues/487

## Does this PR introduce a breaking change?

```
[ ] Yes
[x] No
```

## Motivation
NGXS performs operations inside Angular zone, even if you explicitly call `runOutsideAngular` - e.g. `select` still will be performed inside zone, because of `enterZone` operator.

`outsideZone` is an advanced option for developers who manually control the CD mechanism. E.g. all views in the app are marked as `OnPush` and explicit subscriptions inside views are avoided. All subscribtions are performed using `async` pipe. `async` is an impure pipe and also calls `ViewRef.markForCheck()` under the hood.

### CODE REVIEW

**Comment by splincode** [comment]
@markwhitfeld this is a very important feature, I think it's worth accepting, if in the following PR we can fix the functionality if necessary

File: packages/store/src/internal/dispatcher.ts [hunk 1]
```diff
@@ -96,14 +106,13 @@ export class InternalDispatcher {
...
       .pipe(
         exhaustMap((ctx: ActionContext) => {
-          switch (ctx.status) {
-            case ActionStatus.Successful:
-              return of(this._stateStream.getValue());
-            case ActionStatus.Errored:
-              return throwError(ctx.error);
-            default:
-              return empty();
+          if (ctx.status === ActionStatus.Successful) {
```
Comment (markwhitfeld) [inline packages/store/src/internal/dispatcher.ts]: Why has this been changed from the switch to the if? Please change it back.

File: packages/store/src/operators/wrap.ts [hunk 2]
```diff
@@ -0,0 +1,58 @@
...
+
+class WrapOperator<T> implements Operator<T, T> {
+  constructor(private outsideZone: boolean | null, private zone: NgZone) {}
+
+  public call(subscriber: Subscriber<T>, source: Observable<T>): Subscription {
+    return source.subscribe(new WrapSubscriber(subscriber, this.outsideZone, this.zone));
+  }
+}
+
+class WrapSubscriber<T> extends Subscriber<T> {
```
Comment (markwhitfeld) [inline packages/store/src/operators/wrap.ts]: What is the reason for the vastly different approach to what was used with the enterZone pipe?
This is a lot more code.

File: packages/store/src/internal/dispatcher.ts (hunk 1 above)
Comment (arturovt) [reply packages/store/src/internal/dispatcher.ts]: @markwhitfeld Isn't that the same behavior, except `if-else` doesn't lead to nesting like `switch-case`?

File: packages/store/src/operators/wrap.ts (hunk 2 above)
Comment (arturovt) [reply packages/store/src/operators/wrap.ts]: @markwhitfeld I can re-write `wrap` operator in the same way as `enterZone` was implemented

File: packages/store/src/internal/dispatcher.ts (hunk 1 above)
Comment (splincode) [reply packages/store/src/internal/dispatcher.ts]: https://github.com/ngxs/store/issues/739
https://github.com/ngxs/store/issues/487

in that we can now control the management of the zone

This feature will allow you to disable extra detect changes in angular.

File: packages/store/src/operators/wrap.ts (hunk 2 above)
Comment (splincode) [reply packages/store/src/operators/wrap.ts]: https://github.com/ngxs/store/issues/739
https://github.com/ngxs/store/issues/487

in that we can now control the management of the zone

This feature will allow you to disable extra detect changes in angular.

Comment (splincode) [reply packages/store/src/operators/wrap.ts]: revert by https://github.com/ngxs/store/pull/763

File: packages/store/src/internal/dispatcher.ts (hunk 1 above)
Comment (splincode) [reply packages/store/src/internal/dispatcher.ts]: revert by https://github.com/ngxs/store/pull/763

**Comment by eranshmil** [comment]
The new file, `docs/advanced/options.md`, is not linked in the docs.
Is that on purpose?

**Comment by splincode** [comment]
@eranshmil fixed
https://ngxs.gitbook.io/ngxs/v/master/advanced/options

**Comment by eranshmil** [comment]
Thanks, but I don't think this should be under the `advanced` section.
Maybe inside the installation page.

Btw, I didn't tested it out throughly but when adding `outsideZone: true`, there's an error:
`Error: Right-hand side of 'instanceof' is not an object`

**Comment by splincode** [comment]
@arturovt `Error: Right-hand side of 'instanceof' is not an object`

> Thanks, but I don't think this should be under the advanced section.
Actually quite suitable, as the module setting is advanced parameters.

**Comment by eranshmil** [comment]
I got this error using the old stackblitz template. After upgrading all the deps, it's not happening anymore.

@markwhitfeld We should update the stackblitz template created by Austin.
https://stackblitz.com/edit/ngxs-simple

**Comment by markwhitfeld** [comment]
@eranshmil yes, but I think it should be based off of a github folder so that it is easier to update.
Regarding this zone stuff, I'll be implementing a strategy pattern approach for solving this instead of the current solution

**Pull request merged**